### DISCUSSION TITLE
FastF1 v2.2 Pre-Release Versions

### DISCUSSION THREAD

**Discussion by theOehrly** [theOehrly/Fast-F1#105] [Announcements]
**Version 2.2.0 has now been released.**

....

FastF1 v2.2 is currently under development and intended to be released before the first race weekend of the 2022 season.

Notable changes and new features so far are:

- the implementation of an event schedule
- better caching
- some bug fixes, stability and reliability improvements

A first pre-release version is now available for download below. More pre-release versions will be made available. These versions are still very much under development and should be considered as alpha releases. Further changes and improvements will be made.

Some current functionality has been deprecated in v2.2 to facilitate the implementation of new features and improvements. Deprecated functionality will be removed in v2.3.
A few very limited changes required the immediate removal of previous features without a deprecation grace period.

New features, bug fixes, deprecations and breaking changes are listed in the changelog section of the documentation.

**The documentation for the most current pre-release version is available here:**

Docs: https://theoehrly.github.io/Fast-F1-Pre-Release-Documentation/index.html
Changelog: https://theoehrly.github.io/Fast-F1-Pre-Release-Documentation/changelog.html

**Please leave feedback about these changes below. If you find any bugs in these pre-release versions, please also comment below. The issue section should be reserved for bugs related to stable releases.**

To install the pre-release version, download the zip file below, extract it and install the wheel using `pip install <filename.whl>`.

[v2.2.0_alpha0.zip](https://github.com/theOehrly/Fast-F1/files/8161230/v2.2.0_alpha0.zip)

[v2.2.0_alpha1.zip](https://github.com/theOehrly/Fast-F1/files/8245367/v2.2.0_alpha1.zip)

**Comment by siddharth-tewari** [comment]
Is event schedule being taken from <a href="https://www.formula1.com/en/racing/2022.html">https://www.formula1.com/en/racing/2022.html</a> or <a href="https://api.formula1.com/"> https://api.formula1.com</a> or somewhere else?

**Reply by theOehrly** [reply]
It's kind of a combination of both + data from ergast + some manual additions by me, pre-merged into one dataset so FastF1 does not need to do multiple API requests.

**Reply by siddharth-tewari** [reply]
Ok, that is good

**Comment by walteij** [comment]
Trying to get the test data from the morning session
`>>> session = fastf1.get_session(2022, 'testing', 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "\path	o\python\Python39\site-packages\fastf1\events.py", line 252, in get_session
    raise DeprecationWarning('Accessing test sessions through '
DeprecationWarning: Accessing test sessions through `get_session` has been deprecated!
Use `get_testing_session` instead.`

So I use get_testing_session:
`>>> session = fastf1.get_testing_session(2022, 'testing', 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "\path	o\python\Python39\site-packages\fastf1\events.py", line 288, in get_testing_session
    event = get_testing_event(year, test_number)
  File "\path	o\python\Python39\site-packages\fastf1\events.py", line 343, in get_testing_event
    assert test_number >= 1
TypeError: '>=' not supported between instances of 'str' and 'int'`

using Python 3.9 on Windows 10.

**Reply by walteij** [reply]
I always refused to read properly...
'session = fastf1.get_testing_session(2022, 1, 1)'
Is the solution.

**Comment by theOllieS** [comment]
I'm trying to install the pre release version, however I'm having some issues. 
I have downloaded the [v2.2.0_alpha0.zip](https://github.com/theOehrly/Fast-F1/files/8161230/v2.2.0_alpha0.zip) file. 

However when I try `pip install v2.2.0_alpha0.zip` I get the following error message _ERROR: file:///Users/olliesowden/.spyder-py3/F1%20data/v2.2.0_alpha0.zip does not appear to be a Python project: neither 'setup.py' nor 'pyproject.toml' found._

Am I missing a step here? Thanks

**Reply by walteij** [reply]
Unzip it and use the .whl file, that will work ;)

**Reply by theOllieS** [reply]
Thanks. I have successfully installed the new version. I guess its just a matter of waiting until the data becomes available after the session?

**Reply by walteij** [reply]
You can use the livetiming to collect the data yourself. However, working with the data will only be possible after the session, information about livetiming can be found here: https://theoehrly.github.io/Fast-F1/livetiming.html

**Reply by lolostef** [reply]
> Unzip it and use the .whl file, that will work ;)

Which programm should i use to open the file?
cheers

**Reply by theOehrly** [reply]
Simply use ``pip install <extracted_file.whl>``

It's only zipped because github is very restrictive about the file types that I can append here.

**Reply by lolostef** [reply]
Yeah but on the mac i cant open it resp i need a way to open the file

**Reply by walteij** [reply]
Mac has builtin support for working with zipfiles (according to their support), so just doubleclick the zipfile and move the file inside it to your folder.
https://support.apple.com/guide/mac-help/zip-and-unzip-files-and-folders-on-mac-mchlp2528/mac

**Comment by hotlapmode** [comment]
the data will be available at the end of the day? or is the morning data already available

**Reply by theOehrly** [reply]
It's one single session for morning and afternoon usually. Therefore, data is only available at the end of the day.

You should be able to record the live timing data stream and use the recordings from the morning part of the session to load this part of the session earlier. Using a recording that's cut off at the end should work. (A recording that's cut off at the beginning won't work correctly)

**Reply by hotlapmode** [reply]
thank you. on another note, i can't get the usual codes i used to work on the new version: i tried reading the changelog but i don't know what i have to modify in my code to make it work. Is there anything in particular i should look our for and change in codes compared to last year?

**Reply by theOehrly** [reply]
Are there any errors or warnings that give a hint a what you need to change? There should be if it's intentional. Else it's likely a bug. Can you post the output here?

**Reply by hotlapmode** [reply]
![Screenshot 3_10_2022 5_38_03 PM](https://user-images.githubusercontent.com/67686869/157711164-9a6c87a9-be2e-4342-94f1-a8dd42a6b529.png)
the problem is at the beginning: from the changelog indications i thought that i had to change from [fastf1.get_session()] to [fastf1.get_testing_session()] but it doesn't work.
I'm very new to this so there may be a huge mistake that i don't see

**Reply by theOehrly** [reply]
That looks very much like you still have the "old" stable version installed. You can check with 

```
import fastf1
print(fastf1.__version__)
```

Apart from that, the Bahrain test is test number two. Barcelona was number one although there's no data for it.

**Reply by hotlapmode** [reply]
oh yes that's the problem. i thought i did install the new version, i'll look back and try again. thanks once again

**Reply by siddharth-tewari** [reply]
@theOehrly Is there a way to combine and get testing data for all days? I want to combine all data and make telemetry of fastest lap instead of day-wise

**Comment by theOllieS** [comment]
`session = fastf1.get_testing_session(2022, 2, 1)`
`laps = session.load_laps(with_telemetry=True).pick_accurate()`

This returns _nan_ in the _Team_ and _Driver_ columns when calling `laps` for Albon, Zhou and Fittipaldi, although it does display the correct _Driver Number_

**Reply by mzm-s** [reply]
Yup noticed this here as well, ALB, ZHO and FIT werent classified under their name/team like mentioned above. From session (2022,2,1)

**Reply by professorconst** [reply]
For me get_testing_session brings an error. And get_event_schedule returns only conventional events. Puzzled.

**Reply by dlundgaard** [reply]
What kind of error are you getting with `get_testing_session`? 

If you're getting `fastf1.api.SessionNotAvailableError: No data for this session! Are you sure this session wasn't cancelled?`, that just means the session you are requesting is not (yet) available.

For reference, day 2 of testing will be available shortly using the parameters `get_testing_session(2022, 2, 2)`, but for the moment it results in the previously mentioned error.

**Reply by professorconst** [reply]
I am getting this one:
_ValueError: Test event number 1 does not exist_
But for whatever reason there is another one prior to the above:
_ERROR 	Failed to access primary schedule backend. Falling back to Ergast! Reason: HTTPSConnectionPool(host='raw.githubusercontent.com', port=443):_

**Reply by dlundgaard** [reply]
My guess would be that you are trying to retrieve the ***first*** test of 2022 (the one in Barcelona on which there is no data available), which is what the error message is trying to communicate. 

Thus, `get_testing_session(year=2022, test_number=1, session_number=2)` will result in an error, which is why you will have to retrieve the ***second*** test of 2022 using `get_testing_session(year=2022, test_number=2, session_number=2)`.

**Reply by professorconst** [reply]
I tried with both, but the same error occurs.
Do I understand right that I am supposed to see test events in the list of scheduled sessions for 2022 with `get_event_schedule(2022) `command? So far I can only see conventional events, not testing sessions.

**Reply by dlundgaard** [reply]
As mentioned previously, the data for day 2 (today), are not yet available, so everyone will be getting `SessionNotAvailableError` when trying to retrieve data for day 2 at the moment of writing, although it be available shortly. 

To troubleshoot, you could try getting the data from yesterdays session with `get_testing_session(year=2022, test_number=2, session_number=1)`.

I am not familiar with `get_event_schedule`, so I cannot deny nor confirm whether that is compatible with Pre-Season testing sessions yet.

**Reply by theOehrly** [reply]
@professorconst  ``ERROR Failed to access primary schedule backend `` has nothing to do with requesting the data too early. This is an actual problem somewhere.
It's also the reason why you don't see testing events in the event schedule. The primary schedule backend is maintained and hosted by me. The Ergast database is used as a fallback, but among other things, it does not provide information about testing events. Therefore, the testing event does not exist for FastF1 if it needs to fall back to Ergast and you get the second error.
The error reason seems to be cut off. Can you please provide the full output? I need to figure this out.

**Reply by theOehrly** [reply]
@theOllieS , @mzm-s FastF1 currently requires a hard coded list of drivers to map driver names and teams to the driver number (for testing events only). This list is not up to date in the current release.
Among other things, this will be resolved in the next pre-release version by pulling this data from the api as well.

**Reply by theOllieS** [reply]
@theOehrly thank you for the reply. I managed to figure out a way to get around this until the next update as you have said.

**Reply by professorconst** [reply]
For everyone, thanks. The problem is actually something bigger than just requesting the data too early. The error with primary backend was indicating some connection issues, and when I tried to run the same script on a different network, everything went all well.
I suspect the connection to the data set was somehow blocked, because I can repeat the same error over and over again, while connected to the same problematic network.

**Reply by theOehrly** [reply]
@professorconst The connection to what I call the "primary backend" is really just downloading a few files from a separate GitHub repository. I'm a bit surprised that the connection to GitHub fails, but the connection to Ergast is possible. I was hoping that (mis)using GitHub for this is simple and reliable. You're the only person so far to report a problem with this, but I was hoping that there won't be any problems of this sort.

**Comment by hotlapmode** [comment]
I don't want to open a full discussion for this but is there any way to get the Haas laps after the chequered flag fell for the normal session? I tried yesterday with Magnussen but it gave me no results

edit: today i have the data. all solved

**Comment by markste-in** [comment]
When I try to pick Verstappen I get NaN in the driver column:
Test Session: 2022,2,2

<img width="495" alt="image" src="https://user-images.githubusercontent.com/33337590/158034897-77e20bc2-45a4-4b07-989d-af007fd05c5d.png">

**Reply by walteij** [reply]
yep, see above:
https://github.com/theOehrly/Fast-F1/discussions/105#discussioncomment-2342300

**Reply by dlundgaard** [reply]
As a workaround, and arguably a safer solution in the long run, you can select the driver by his three-letter FIA abbreviation, in your case `pick_driver("VER")`.

**Reply by markste-in** [reply]
Yes that is what I normally do in the stable release but doesn't work in the current alpha since the mapping is not ok
<img width="530" alt="image" src="https://user-images.githubusercontent.com/33337590/158035115-25a1fd93-844f-4eed-9396-4c0f712f8d60.png">

**Reply by walteij** [reply]
Workaround is to use the drivernumber
`laps.pick.driver("VER")["DriverNumber"]`

**Comment by glokta1** [comment]
@theOehrly Is the package on track to be released before the first race weekend?

**Reply by theOehrly** [reply]
I think I'll get everything finished in time. I'm running a bit out of time but I think it'll work out.

**Comment by Majd-ui** [comment]
`ValueError: Must pass 2-d input. shape=(1, 81, 25)`

I get this error when I Try to sort out a list of laps by Time lap value

Im a beginner

**Reply by theOehrly** [reply]
Please post a minimal code example of what you are trying to do.

**Reply by Majd-ui** [reply]
Thank you for your response, I actually figured it out.
I was trying to do ur tutorial of extracting every driver quali lap but I changed it, instead I wanted to extract one driver laptimes in testing and list it. For some reason I thought I needed to do a list() for that.  But you can help with an other thing, do you know wich plot to use to list the LapTiming?

**Reply by theOehrly** [reply]
Can you explain more clearly what you want to do? Also, please open a separate new discussion, as this is not the right place for  general questions.

**Comment by theOehrly** [comment]
Next pre-release version is available. I updated the original post above. Stable will follow later this week, likely only on Thursday. Sorry for being a bit late.

Main changes include improved support for loading driver lists and session results. This also fixes the issue with missing driver abbreviations in testing events that was present in the last release.

If you already work with the event schedule object directly, here's a short warning: Currently, the dataframe columns are named 'event_name', 'location' and similar. This will be changed to 'EventName', 'Location', ... to match the naming convention that was already used for the Laps object. There will be absolutely no backwards compatibility from stable to alpha! Sorry, this was not well thought through in the first place, but I want to change it now before doing a stable release with this.

**Comment by harningle** [comment]
I'm little bit confused about tyre compounds... I guess the raw live timing data could messy. The `TimingAppData.jsonStream` seems to have five compounds: {"SOFT", "MEDIUM", "HARD", "UNKNOWN", "TEST_KNOWN"}. But if you look at day 3 (2022-03-12), it says Ver got 1:31.720 with "SOFT" tyre. Lec also had 1:32.415 with "SOFT" tyre. However, according to [Sky Sports F1](https://www.skysports.com/f1/news/12433/12564276/f1-testing-bahrain-day-three-red-bulls-max-verstappen-ahead-of-ferrari-as-mercedes-trail-2022-rivals), the "SOFT" for Ver is C5, while the same "SOFT" for Lec is C4.

**Reply by harningle** [reply]
I cross-validate the data from live timing and other sources. My guess is that both C4 and C5 are labeled as SOFT in live timing, and C3 is MEDIUM, C4 and C5 are HARD. UNKNOWN is C4 and TEST_UNKNOWN is C2. It's somewhat frustrating that live timing doesn't provide the original C1 to C5 info.

**Reply by theOehrly** [reply]
The API usually only supports SOFT, MEDIUM, HARD, INTERMEDIATE, WET without taking into account C1 to C5. This is sufficient for normal race weekends. Apparently the compounds available for testing were the following:

![grafik](https://user-images.githubusercontent.com/23384863/158208241-ed63f5eb-a4a8-43d0-bd91-8a7934b38e78.png)

C1 to C5 + one prototype + plus intermediate and wet tyre. Therefore, it makes sense that they can not be differentiated properly in the API which is unfortunate but there's not much we can do about it. I haven't tried to figure out which of theses compounds is represented by which name in the API.

**Reply by harningle** [reply]
Yes it's like a shortcoming of the API. But anyway. Thanks for your great package!

**Comment by OxfordComma** [comment]
After updating to alpha1, the "Team" values are blank for the laps in Practice 1:
<img width="1360" alt="image" src="https://user-images.githubusercontent.com/8374102/158476462-f1d77c54-f350-4684-84df-e808dc409c91.png">
But not Practice 2 or 3:
<img width="1362" alt="image" src="https://user-images.githubusercontent.com/8374102/158476547-92f1cc8c-8e9d-4156-97ed-1a3a66316111.png">
<img width="1364" alt="image" src="https://user-images.githubusercontent.com/8374102/158476584-81fb1999-4725-4d89-920f-221f417d498a.png">

**Reply by theOehrly** [reply]
Will be fixed, thanks for the info.

**Comment by DavideMagno** [comment]
Any news on the official release of 2.2?

**Reply by theOehrly** [reply]
Done! 😉

**Comment by theOehrly** [comment]
v2.2.0 stable is now released and can be installed through pip as usual.

**Reply by walteij** [reply]
Grabbing TheOneScript to see if everything is working properly ;)
........

Only some small name changes since the alpha release, nothing so serious it could not be fixed in 30 seconds.

**Comment by silvertales** [comment]
I'm fairly new to the world of Python, and I'm still learning so obviously I don't know everything. I'd like to see how this package works but I'm very confused and lost. I've come across many issues but here is the very first obstacle. I'd appreciate help, and an explanation in simple terms. 

![issue](https://user-images.githubusercontent.com/102412335/161207441-2a1421d3-0787-47d3-9712-7ffac7abebf6.jpg)

**Reply by walteij** [reply]
In line 7 there's `('path/to/folder/for/cache')`
That needs to point to an existing folder, so if you create a folder named cache in your current working directory change that bit to `('./cache')`
If your cache folder is in C:\Temp change that bit to `('C:\temp\cache')`

**Reply by silvertales** [reply]
Do you mean that the path needs to be the FastF1 folder that i downloaded from the code section? I searched it all and did not find any folder or file named cache.

**Reply by walteij** [reply]
No, you need to create the cache folder yourself.

**Reply by silvertales** [reply]
So I go into My PC and just create a folder called cache? what am I supposed to put inside that folder?

**Reply by walteij** [reply]
Nothing. When you create the folder, the module will use that folder to save the data to, so you don't need to download it again when you run the same script again, of when you run another script for the same race/event.

**Reply by markste-in** [reply]
make it easy and just use 'cache' for the path and create an empty folder in the same place where ur python file is.
`fastf1.Cache.enable_cache('cache')`

**Reply by silvertales** [reply]
was this the result I was supposed to get?

![test2](https://user-images.githubusercontent.com/102412335/161213183-5f300a89-b58f-4e7d-8035-72e6cbb97343.jpg)

**Reply by markste-in** [reply]
Yeah. You are not doing anything yet. You just loaded the lap data into `lec` and `ham`

**Reply by walteij** [reply]
Well, since this is only selecting the fastest laps for Leclerc and Hamilton yes.
Now you need to decide what you want to do with these fastest laps.
If you do `lec()` or `ham()` it should show you the data for their fastest lap, that you can use to plot graphs.

**Reply by silvertales** [reply]
I know my questions are way beyond dumb, but where should I enter the lec() or ham()

**Reply by markste-in** [reply]
just do a 
```
print(lec)
print(ham)
```
if you just want to print out the data

**Reply by silvertales** [reply]
still getting a 0, unless I'm doing something wrong. 
![test3](https://user-images.githubusercontent.com/102412335/161215127-87f1a4d8-6b66-40c0-a8b1-11129f68c8b5.png)

**Reply by markste-in** [reply]
no...
you should see something like this:

```
core           INFO     Loading data for Turkish Grand Prix - Race [v2.2.2]
api            INFO     Using cached data for driver_info
api            INFO     Using cached data for timing_data
api            INFO     Using cached data for timing_app_data
core           INFO     Processing timing data...
api            INFO     Using cached data for session_status_data
api            INFO     Using cached data for track_status_data
api            INFO     Using cached data for car_data
api            INFO     Using cached data for position_data
api            INFO     Using cached data for weather_data
core           INFO     Finished loading data for 20 drivers: ['44', '11', '5', '16', '55', '33', '23', '4', '18', '3', '31', '26', '10', '77', '7', '63', '20', '8', '6', '99']
                     Time DriverNumber                LapTime  LapNumber  Stint  ...      Team Driver TrackStatus IsAccurate            LapStartDate
0  0 days 00:35:38.442000           44                    NaT          1      1  ...  Mercedes    HAM           2      False 2020-11-15 10:13:26.827
1  0 days 00:37:38.323000           44 0 days 00:01:59.881000          2      1  ...  Mercedes    HAM           2       True 2020-11-15 10:15:38.944
2  0 days 00:39:34.982000           44 0 days 00:01:56.659000          3      1  ...  Mercedes    HAM           1       True 2020-11-15 10:17:38.825
3         0 days 00:41:30           44 0 days 00:01:55.018000          4      1  ...  Mercedes    HAM           1       True 2020-11-15 10:19:35.484
4  0 days 00:43:24.724000           44 0 days 00:01:54.724000          5      1  ...  Mercedes    HAM           1       True 2020-11-15 10:21:30.502
5  0 days 00:45:18.007000           44 0 days 00:01:53.283000          6      1  ...  Mercedes    HAM           1       True 2020-11-15 10:23:25.226
6  0 days 00:47:10.569000           44 0 days 00:01:52.562000          7      1  ...  Mercedes    HAM           1       True 2020-11-15 10:25:18.509
7  0 days 00:49:04.970000           44 0 days 00:01:54.401000          8      1  ...  Mercedes    HAM           1      False 2020-11-15 10:27:11.071
8  0 days 00:51:13.827000           44 0 days 00:02:08.857000          9      2  ...  Mercedes    HAM           1      False 2020-11-15 10:29:05.472
9  0 days 00:52:59.699000           44 0 days 00:01:45.872000         10      2  ...  Mercedes    HAM           1       True 2020-11-15 10:31:14.329
10 0 days 00:54:48.908000           44 0 days 00:01:49.209000         11      2  ...  Mercedes    HAM           1       True 2020-11-15 10:33:00.201
11 0 days 00:56:39.147000           44 0 days 00:01:50.239000         12      2  ...  Mercedes    HAM           1       True 2020-11-15 10:34:49.410
12 0 days 00:58:33.618000           44 0 days 00:01:54.471000         13      2  ...  Mercedes    HAM          26      False 2020-11-15 10:36:39.649
13 0 days 01:00:40.272000           44 0 days 00:02:06.654000         14      2  ...  Mercedes    HAM           6      False 2020-11-15 10:38:34.120
14 0 days 01:02:35.625000           44 0 days 00:01:55.353000         15      2  ...  Mercedes    HAM          67      False 2020-11-15 10:40:40.774
15 0 days 01:04:22.818000           44 0 days 00:01:47.193000         16      2  ...  Mercedes    HAM           1       True 2020-11-15 10:42:36.127
16 0 days 01:06:08.862000           44 0 days 00:01:46.044000         17      2  ...  Mercedes    HAM           2       True 2020-11-15 10:44:23.320
17 0 days 01:07:54.585000           44 0 days 00:01:45.723000         18      2  ...  Mercedes    HAM           2       True 2020-11-15 10:46:09.364
18 0 days 01:09:38.701000           44 0 days 00:01:44.116000         19      2  ...  Mercedes    HAM           1       True 2020-11-15 10:47:55.087
19 0 days 01:11:22.649000           44 0 days 00:01:43.948000         20      2  ...  Mercedes    HAM           1       True 2020-11-15 10:49:39.203
20 0 days 01:13:05.558000           44 0 days 00:01:42.909000         21      2  ...  Mercedes    HAM           2       True 2020-11-15 10:51:23.151
21 0 days 01:14:49.151000           44 0 days 00:01:43.593000         22      2  ...  Mercedes    HAM           1       True 2020-11-15 10:53:06.060
22 0 days 01:16:33.287000           44 0 days 00:01:44.136000         23      2  ...  Mercedes    HAM           1       True 2020-11-15 10:54:49.653
23 0 days 01:18:17.067000           44 0 days 00:01:43.780000         24      2  ...  Mercedes    HAM           1       True 2020-11-15 10:56:33.789
24 0 days 01:20:00.203000           44 0 days 00:01:43.136000         25      2  ...  Mercedes    HAM           1       True 2020-11-15 10:58:17.569
25 0 days 01:21:42.295000           44 0 days 00:01:42.092000         26      2  ...  Mercedes    HAM           1       True 2020-11-15 11:00:00.705
26 0 days 01:23:25.006000           44 0 days 00:01:42.711000         27      2  ...  Mercedes    HAM           1       True 2020-11-15 11:01:42.797
27 0 days 01:25:08.001000           44 0 days 00:01:42.995000         28      2  ...  Mercedes    HAM           1       True 2020-11-15 11:03:25.508
28 0 days 01:26:50.476000           44 0 days 00:01:42.475000         29      2  ...  Mercedes    HAM           1       True 2020-11-15 11:05:08.503
29 0 days 01:28:33.714000           44 0 days 00:01:43.238000         30      2  ...  Mercedes    HAM           1       True 2020-11-15 11:06:50.978
30 0 days 01:30:17.497000           44 0 days 00:01:43.783000         31      2  ...  Mercedes    HAM           1       True 2020-11-15 11:08:34.216
31 0 days 01:32:01.852000           44 0 days 00:01:44.355000         32      2  ...  Mercedes    HAM           1       True 2020-11-15 11:10:17.999
32 0 days 01:33:44.170000           44 0 days 00:01:42.318000         33      2  ...  Mercedes    HAM           1       True 2020-11-15 11:12:02.354
33 0 days 01:35:25.834000           44 0 days 00:01:41.664000         34      2  ...  Mercedes    HAM           2       True 2020-11-15 11:13:44.672
34 0 days 01:37:07.492000           44 0 days 00:01:41.658000         35      2  ...  Mercedes    HAM           1       True 2020-11-15 11:15:26.336
35 0 days 01:38:50.098000           44 0 days 00:01:42.606000         36      2  ...  Mercedes    HAM           1       True 2020-11-15 11:17:07.994
36 0 days 01:40:32.187000           44 0 days 00:01:42.089000         37      2  ...  Mercedes    HAM           1       True 2020-11-15 11:18:50.600
37 0 days 01:42:14.143000           44 0 days 00:01:41.956000         38      2  ...  Mercedes    HAM           1       True 2020-11-15 11:20:32.689
38 0 days 01:43:56.009000           44 0 days 00:01:41.866000         39      2  ...  Mercedes    HAM           2       True 2020-11-15 11:22:14.645
39 0 days 01:45:36.646000           44 0 days 00:01:40.637000         40      2  ...  Mercedes    HAM           2       True 2020-11-15 11:23:56.511
40 0 days 01:47:16.907000           44 0 days 00:01:40.261000         41      2  ...  Mercedes    HAM           2       True 2020-11-15 11:25:37.148
41 0 days 01:48:56.906000           44 0 days 00:01:39.999000         42      2  ...  Mercedes    HAM           1       True 2020-11-15 11:27:17.409
42 0 days 01:50:37.395000           44 0 days 00:01:40.489000         43      2  ...  Mercedes    HAM           1       True 2020-11-15 11:28:57.408
43 0 days 01:52:17.986000           44 0 days 00:01:40.591000         44      2  ...  Mercedes    HAM           1       True 2020-11-15 11:30:37.897
44 0 days 01:53:58.965000           44 0 days 00:01:40.979000         45      2  ...  Mercedes    HAM           1       True 2020-11-15 11:32:18.488
45 0 days 01:55:41.133000           44 0 days 00:01:42.168000         46      2  ...  Mercedes    HAM           1       True 2020-11-15 11:33:59.467
46 0 days 01:57:21.901000           44 0 days 00:01:40.768000         47      2  ...  Mercedes    HAM           1       True 2020-11-15 11:35:41.635
47 0 days 01:59:03.286000           44 0 days 00:01:41.385000         48      2  ...  Mercedes    HAM           1       True 2020-11-15 11:37:22.403
48 0 days 02:00:44.262000           44 0 days 00:01:40.976000         49      2  ...  Mercedes    HAM           2       True 2020-11-15 11:39:03.788
49 0 days 02:02:24.745000           44 0 days 00:01:40.483000         50      2  ...  Mercedes    HAM           1       True 2020-11-15 11:40:44.764
50 0 days 02:04:05.180000           44 0 days 00:01:40.435000         51      2  ...  Mercedes    HAM           2       True 2020-11-15 11:42:25.247
51 0 days 02:05:45.055000           44 0 days 00:01:39.875000         52      2  ...  Mercedes    HAM           1       True 2020-11-15 11:44:05.682
52 0 days 02:07:25.420000           44 0 days 00:01:40.365000         53      2  ...  Mercedes    HAM           1       True 2020-11-15 11:45:45.557
53 0 days 02:09:05.886000           44 0 days 00:01:40.466000         54      2  ...  Mercedes    HAM           1       True 2020-11-15 11:47:25.922
54 0 days 02:10:46.618000           44 0 days 00:01:40.732000         55      2  ...  Mercedes    HAM           1       True 2020-11-15 11:49:06.388
55 0 days 02:12:26.031000           44 0 days 00:01:39.413000         56      2  ...  Mercedes    HAM           1       True 2020-11-15 11:50:47.120
56 0 days 02:14:05.603000           44 0 days 00:01:39.572000         57      2  ...  Mercedes    HAM           1       True 2020-11-15 11:52:26.533
57 0 days 02:15:45.851000           44 0 days 00:01:40.248000         58      2  ...  Mercedes    HAM           1       True 2020-11-15 11:54:06.105

[58 rows x 27 columns]

Process finished with exit code 0
```

Be careful with your cache folder. You have whitespaces in your path. I am not sure how python handles that on Windows. Maybe just create an empty "cache" folder in your current folder where you also have your python file

**Reply by markste-in** [reply]
When you use Windows you should also use double backslashes for your path
'D:\\\Erfan Backup\\\FastF1\\\cache'

**Reply by walteij** [reply]
forward slashes will work as well (at least, they do for me :) )

**Reply by silvertales** [reply]
If you mean by having an empty folder in the same folder as the Fastf1 folder, then it is already there. I also changed the path to the one you mentioned and tried forward slashes too. I keep getting this error. 

![test3](https://user-images.githubusercontent.com/102412335/161218279-04094d13-fc62-4d5f-b386-dd847e9c94d5.png)
![test4](https://user-images.githubusercontent.com/102412335/161218314-0a61a811-feb1-430e-b586-91887a43496c.jpg)

**Reply by silvertales** [reply]
some progress but...is there an issue still?
![test5](https://user-images.githubusercontent.com/102412335/161218641-6cbf6485-0650-4834-ab76-ddcce27f334b.jpg)

**Comment by silvertales** [comment]
I think I got it!
![test5](https://user-images.githubusercontent.com/102412335/161218801-a0260996-6ddf-4141-b7b4-59775c4f8cae.jpg)

**Reply by markste-in** [reply]
yes this is the expected outcome for ur code

**Reply by markste-in** [reply]
if you want to test plotting you can create something simple
```
plt.plot(lec["SpeedST"])
plt.show()
```
![myplot](https://user-images.githubusercontent.com/33337590/161237400-a9912ee1-290b-4348-bae3-d0614a0229f9.png)

**Comment by silvertales** [comment]
Does python show telemetry? I've seen some people say they can gather data from python and make telemetry, so I was wondering if thats a feature python can has, or if the data is collected from python and taken elsewhere to create a telemetry chart.

**Reply by walteij** [reply]
Python itself can create charts from the data it collects from fastf1 using modules like matplotlib. You should check the [examples page](https://theoehrly.github.io/Fast-F1/examples_gallery/index.html) for some explanation.

**Reply by markste-in** [reply]
It actually works like this:
1. The data is captured from a live stream during the event.
2. After the event it get processed and pushed to a database ([Ergast DB](http://ergast.com/mrd/))
3. From there on you can either access Ergast yourself (if you know how to use SQL) or you use FastF1. 
FastF1 is just a package written in python, to simplify access to telemetry data. It also has the ability to (re)calculate certain data (like Distance).

Python itself is just a programming language and has nothing to do with telemetry. It gives you the capability to do your own calculations and plot those results (as shown in the examples). The advantage of python is its simplicity, the giant amount of available packages and its superb community support.

**Reply by silvertales** [reply]
I've spent a few hours today testing some, and I've been getting good result. Of course, I'm a beginner and this is too advanced for me but I've always been good at learning by doing more practical work. I love data analysis and telemetry, so I'd really like to get into Python professionally and play around with F1 data. I'll still go back to learning so I can truly master Python without asking such noob questions - which are normal for a beginner but still. 

I appreciate the help and guidance from you both, and I'll definitely ask again if I need help. 

Side question: I'm learning Python from a youtube video at the moment. Would you say its good enough to get started?

**Reply by walteij** [reply]
youtube video's can be a good source to learn from, but do not underestimate the power of a good book and on-the-job training.
I only have some basic powershell skills, and started with python a couple of days (maybe 2 weeks) after the season ending 2021.
Took a proper book about python, did the excersices, made some serious errors, and learned from that.
You need to know the python fundamentals, it's not really that hard, and if you have them and read the error reports your script sends you, you should improve quick enough.

I even ventured into creating a discord bot that generates telemetry plots per request. And so far it seems to be doing the trick, but I still have much to learn as well (which I do by making mistakes).

**Reply by markste-in** [reply]
I'd say stick with youtube. That is more than good enough. 
I never learned a programming language from a book.

Just make sure u actually code. Some people just read and watch a lot about what they wanna do and forget to actually do it

**Reply by silvertales** [reply]
I read in the documentation that we can record live data during a session. I was planning on waking up early for FP1 and while watching, I could record the live data. I wanted to be sure as to what I have to do; should I simply open cmd and type 'python -m fastf1.livetiming save saved_data.txt'? I tried it and nothing got a invalid syntax error.  

Is that exactly what needs to be typed or do we have to add a path? I haven't been able to continue my Python lessons since I had a busy week, so I'm still a little lost. Hoping to get back on track this week.

**Reply by theOehrly** [reply]
Yes, that is the command that you have to use. Either you don't have FastF1 installed or you don't have python installed. Or your shell/cmd can't find your python installation.
